utilities/joint-trajectory-player/src/Module.cpp

```diff
@@ -0,0 +1,291 @@
+        }
+
+        Eigen::VectorXd vector;
+        vector.resize(num_fields);
+        std::size_t found_lines = 0;
+        for (auto line : istrm_strings)
+        {
+            std::size_t found_fields = 0;
+            std::string number_str;
+            std::istringstream iss(line);
+
+            while (iss >> number_str)
+            {
+                vector(found_fields) = std::stod(number_str);
+                found_fields++;
+            }
+            if (num_fields != found_fields)
+            {
+                std::cout << "[Module::readStateFromFile] Malformed input file " << filename
+                          << std::endl;
+                std::cout << "[Module::readStateFromFile] Expected " << num_fields
+                          << " columns, found " << found_fields << std::endl;
+
+                return std::make_pair(false, data);
+            }
+            data.push_back(vector);
+            found_lines++;
+        }
+
+        return std::make_pair(true, data);
+    }
+}
+
+bool Module::configure(yarp::os::ResourceFinder& rf)
+{
+    auto parametersHandler = std::make_shared<ParametersHandler::YarpImplementation>(rf);
+
+    std::string name;
+    if (!parametersHandler->getParameter("name", name))
+        return false;
+    this->setName(name.c_str());
+
+    if (!parametersHandler->getParameter("sampling_time", m_dT))
+        return false;
+
+    std::string trajectoryFile;
+    if (!parametersHandler->getParameter("trajectory_file", trajectoryFile))
+        return false;
+
+    this->createPolydriver(parametersHandler);
+    this->initializeRobotControl(parametersHandler);
+    this->instantiateSensorBridge(parametersHandler);
+
+    m_currentJointPos.resize(m_robotControl.getJointList().size());
+
+    m_qDesired.clear();
+    auto data = readStateFromFile(trajectoryFile, m_numOfJoints);
+    if (!data.first)
+    {
+        return false;
+    }
+    m_qDesired = data.second;
+
+    // check if vector is not initialized
+    if (m_qDesired.empty())
+    {
+        std::cerr << "[Module::configure] Cannot advance empty reference signals." << std::endl;
+        return false;
+    }
+
+    std::cout << "[Module::configure] Starting the experiment." << std::endl;
+
+    // Reach the first position of the desired trajectory in position control
+    m_robotControl.setReferences(m_qDesired.front(),
+                                 RobotInterface::IRobotControl::ControlMode::Position);
+
+    return true;
+}
+
+bool Module::advanceReferenceSignals()
+{
+    m_qDesired.pop_front();
+
+    // check if the vector is empty. If true the trajectory is ended
+    if (m_qDesired.empty())
+    {
+        return false;
+    }
+
+    return true;
+}
+
+bool Module::updateModule()
+{
+    bool isMotionDone;
+    bool isMimeExpired;
+    std::vector<std::pair<std::string, double>> jointlist;
+
+    m_robotControl.checkMotionDone(isMotionDone, isMimeExpired, jointlist);
+
+    if (isMotionDone)
```

Comment:
This function was not meant to be used in this way. What I did in the past is to create a state machine and use `checkMotionDone` as a trigger to change the state. We can discuss this offline